File: scratch/analyze_2025.js

const fs = require('fs');
const readline = require('readline');

function parseCSVLine(line) {
    const result = [];
    let startValueIndex = 0;
    let inQuotes = false;
    for (let i = 0; i < line.length; i++) {
        if (line[i] === '"') {
            inQuotes = !inQuotes;
        } else if (line[i] === ',' && !inQuotes) {
            result.push(line.substring(startValueIndex, i).replace(/^"|"$/g, '').trim());
            startValueIndex = i + 1;
        }
    }
    result.push(line.substring(startValueIndex).replace(/^"|"$/g, '').trim());
    return result;
}

async function analyze() {
    const filePath = 'c:\\xampp\\htdocs\\matadata\\raw\\realisasi2025alldept.csv';
    const fileStream = fs.createReadStream(filePath);

    const rl = readline.createInterface({
        input: fileStream,
        crlfDelay: Infinity
    });

    let header = null;
    const vendorFreq = {};
    const vendorValue = {};
    const skpdValue = {};
    const outsideSystemVendors = {};
    const packetSplittingCandidates = {};
    const foodItems = [];

    for await (const line of rl) {
        if (!line.trim()) continue;
        const cleanParts = parseCSVLine(line);

        if (!header) {
            header = cleanParts;
            continue;
        }

        const getIdx = (name) => header.indexOf(name);
        
        const skpd = cleanParts[getIdx('Nama Satuan Kerja')] || 'UNKNOWN';
        const vendor = (cleanParts[getIdx('Nama Penyedia')] || 'UNKNOWN').toUpperCase();
        const valueStr = cleanParts[getIdx('Total Nilai (Rp)')] || '0';
        const value = parseFloat(valueStr.replace(/,/g, '')) || 0;
        const status = cleanParts[getIdx('Status Paket')] || '';
        const package = cleanParts[getIdx('Nama Paket')] || '';
        const method = cleanParts[getIdx('Metode Pengadaan')] || '';

        // stats
        vendorFreq[vendor] = (vendorFreq[vendor] || 0) + 1;
        vendorValue[vendor] = (vendorValue[vendor] || 0) + value;
        skpdValue[skpd] = (skpdValue[skpd] || 0) + value;

        if (status === 'PAYMENT OUTSIDE SYSTEM') {
            outsideSystemVendors[vendor] = (outsideSystemVendors[vendor] || 0) + 1;
        }

        if (value >= 195000000 && value <= 200000000 && method === 'Pengadaan Langsung') {
            const key = `${vendor} -> ${skpd}`;
            if (!packetSplittingCandidates[key]) packetSplittingCandidates[key] = [];
            packetSplittingCandidates[key].push({ package, value });
        }

        if (package.match(/MAKAN|MINUM|SNACK|NASI|JAMUAN/i)) {
            foodItems.push({ vendor, skpd, package, value });
        }
    }

    const results = {
        topVendorsFreq: Object.entries(vendorFreq).sort((a,b) => b[1] - a[1]).slice(0, 50),
        topVendorsValue: Object.entries(vendorValue).sort((a,b) => b[1] - a[1]).slice(0, 30),
        topSkpdValue: Object.entries(skpdValue).sort((a,b) => b[1] - a[1]).slice(0, 30),
        topOutsideSystem: Object.entries(outsideSystemVendors).sort((a,b) => b[1] - a[1]).slice(0, 30),
        splittingCandidates: Object.entries(packetSplittingCandidates).filter(e => e[1].length > 1).sort((a,b) => b[1].length - a[1].length).slice(0, 20),
        foodSpendingTotal: foodItems.reduce((acc, curr) => acc + curr.value, 0),
        foodTopVendors: Object.entries(foodItems.reduce((acc, curr) => {
            acc[curr.vendor] = (acc[curr.vendor] || 0) + curr.value;
            return acc;
        }, {})).sort((a,b) => b[1] - a[1]).slice(0, 30),
        strangeVendorsCheck: Object.entries(vendorFreq).filter(([name]) => name.includes('ALHAMDULILLAH') || name === 'BADUL' || name === 'OREN' || name === 'KENARA PELITA RAYA').sort((a,b) => b[1]-a[1])
    };

    fs.writeFileSync('c:\\xampp\\htdocs\\matadata\\scratch\\deep_analysis_2025.json', JSON.stringify(results, null, 4));
    console.log('Analysis finished.');
}

analyze();

File: scratch/analyze_2025.py

import pandas as pd
import json

# Set path to the CSV file
file_path = r'c:\xampp\htdocs\matadata\raw\realisasi2025alldept.csv'

# Load the data
# Using low_memory=False to avoid DtypeWarning and handling possible encoding issues
try:
    df = pd.read_csv(file_path, encoding='utf-8')
except UnicodeDecodeError:
    df = pd.read_csv(file_path, encoding='latin-1')

# Data Cleaning
df['Total Nilai (Rp)'] = pd.to_numeric(df['Total Nilai (Rp)'], errors='coerce')
df['Nama Penyedia'] = df['Nama Penyedia'].str.upper().str.strip()
df['Nama Satuan Kerja'] = df['Nama Satuan Kerja'].str.strip()

# 1. Top Vendors by Frequency (Most active)
top_vendors_freq = df['Nama Penyedia'].value_counts().head(20).to_dict()

# 2. Top Vendors by Total Value (Richest)
top_vendors_value = df.groupby('Nama Penyedia')['Total Nilai (Rp)'].sum().sort_values(ascending=False).head(20).to_dict()

# 3. Top SKPD by Makanan & Minuman spending
makan_minum_df = df[df['Nama Paket'].str.contains('MAKAN|MINUM|SNACK|NASI', case=False, na=False)]
top_skpd_makan = makan_minum_df.groupby('Nama Satuan Kerja')['Total Nilai (Rp)'].sum().sort_values(ascending=False).head(15).to_dict()

# 4. Analyze "ALHAMDULILLAH" reaches
alhamdulillah_reach = df[df['Nama Penyedia'] == 'ALHAMDULILLAH'].groupby('Nama Satuan Kerja')['Total Nilai (Rp)'].sum().to_dict()

# 5. Price Variation for "NASI BOX" / "SNACK"
# Let's look for packages like "NASI BOX" and see the price distribution if available, 
# but since the values are total values, we might need a more granular look or just see total per packet.
nasi_box_packets = df[df['Nama Paket'].str.contains('NASI BOX|PAKET MAKAN', case=False, na=False)][['Nama Paket', 'Nama Satuan Kerja', 'Total Nilai (Rp)']].head(20).to_dict(orient='records')

# 6. Method Distribution
method_dist = df['Metode Pengadaan'].value_counts().head(10).to_dict()

# 7. Status Distribution
status_dist = df['Status Paket'].value_counts().head(10).to_dict()

# 8. Anomaly: E-Purchasing vs Payment Outside System for top vendors
outside_system_vendors = df[df['Status Paket'] == 'PAYMENT OUTSIDE SYSTEM']['Nama Penyedia'].value_counts().head(10).to_dict()

# Output the results
analysis_results = {
    "top_vendors_by_frequency": top_vendors_freq,
    "top_vendors_by_value": top_vendors_value,
    "top_skpd_spending_on_food": top_skpd_makan,
    "alhamdulillah_vendor_analysis": alhamdulillah_reach,
    "nasi_box_sample": nasi_box_packets,
    "procurement_methods": method_dist,
    "packet_statuses": status_dist,
    "top_vendors_outside_system": outside_system_vendors
}

with open(r'c:\xampp\htdocs\matadata\scratch\analysis_2025_deep.json', 'w') as f:
    json.dump(analysis_results, f, indent=4)

print("Deep Analysis Complete. Results saved to scratch/analysis_2025_deep.json")
